### PR TITLE
feat: Add support to enable FullTimeStamp in logging

### LINKED ISSUE #8523: Ability to enable FullTimeStamp option in logs based in environment variable
# Summary

Be able to enable or disable `FullTimeStamp` in `logrus.Formatter` used for logs.

# Motivation

It can be very useful to have logs that have the `FullTimeStamp` option enabled or not depending on whether an environment variable `ARGOCD_LOG_FORMAT_ENABLE_TIMESTAMP` has the value 1 or not, just as it is possible to indicate the log format through the `ARGOCD_LOG_FORMAT` environment variable between text or json.

# Proposal

It would be necessary to change the following files:
- Add a const to indicate the environment variable name used to enable `FullTimeStamp` option in logs, for that purpose is necessary modify the file `common/common.go` as exposed below:
	
```
// Environment variables for tuning and debugging Argo CD
const (
	// EnvVarSSODebug is an environment variable to enable additional OAuth debugging in the API server
	EnvVarSSODebug = "ARGOCD_SSO_DEBUG"
	// EnvVarRBACDebug is an environment variable to enable additional RBAC debugging in the API server
	EnvVarRBACDebug = "ARGOCD_RBAC_DEBUG"
	// Overrides the location where SSH known hosts for repo access data is stored
	EnvVarSSHDataPath = "ARGOCD_SSH_DATA_PATH"
	// Overrides the location where TLS certificate for repo access data is stored
	EnvVarTLSDataPath = "ARGOCD_TLS_DATA_PATH"
	// Specifies number of git remote operations attempts count
	EnvGitAttemptsCount = "ARGOCD_GIT_ATTEMPTS_COUNT"
	// Specifices max duration of git remote operation retry
	EnvGitRetryMaxDuration = "ARGOCD_GIT_RETRY_MAX_DURATION"
	// Specifies duration of git remote operation retry
	EnvGitRetryDuration = "ARGOCD_GIT_RETRY_DURATION"
	// Specifies fator of git remote operation retry
	EnvGitRetryFactor = "ARGOCD_GIT_RETRY_FACTOR"
	// Overrides git submodule support, true by default
	EnvGitSubmoduleEnabled = "ARGOCD_GIT_MODULES_ENABLED"
	// EnvGnuPGHome is the path to ArgoCD's GnuPG keyring for signature verification
	EnvGnuPGHome = "ARGOCD_GNUPGHOME"
	// EnvWatchAPIBufferSize is the buffer size used to transfer K8S watch events to watch API consumer
	EnvWatchAPIBufferSize = "ARGOCD_WATCH_API_BUFFER_SIZE"
	// EnvPauseGenerationAfterFailedAttempts will pause manifest generation after the specified number of failed generation attempts
	EnvPauseGenerationAfterFailedAttempts = "ARGOCD_PAUSE_GEN_AFTER_FAILED_ATTEMPTS"
	// EnvPauseGenerationMinutes pauses manifest generation for the specified number of minutes, after sufficient manifest generation failures
	EnvPauseGenerationMinutes = "ARGOCD_PAUSE_GEN_MINUTES"
	// EnvPauseGenerationRequests pauses manifest generation for the specified number of requests, after sufficient manifest generation failures
	EnvPauseGenerationRequests = "ARGOCD_PAUSE_GEN_REQUESTS"
	// EnvControllerReplicas is the number of controller replicas
	EnvControllerReplicas = "ARGOCD_CONTROLLER_REPLICAS"
	// EnvControllerShard is the shard number that should be handled by controller
	EnvControllerShard = "ARGOCD_CONTROLLER_SHARD"
	// EnvEnableGRPCTimeHistogramEnv enables gRPC metrics collection
	EnvEnableGRPCTimeHistogramEnv = "ARGOCD_ENABLE_GRPC_TIME_HISTOGRAM"
	// EnvGithubAppCredsExpirationDuration controls the caching of Github app credentials. This value is in minutes (default: 60)
	EnvGithubAppCredsExpirationDuration = "ARGOCD_GITHUB_APP_CREDS_EXPIRATION_DURATION"
	// EnvHelmIndexCacheDuration controls how the helm repository index file is cached for (default: 0)
	EnvHelmIndexCacheDuration = "ARGOCD_HELM_INDEX_CACHE_DURATION"
	// EnvRepoServerConfigPath allows to override the configuration path for repo server
	EnvAppConfigPath = "ARGOCD_APP_CONF_PATH"
	// EnvLogFormat log format that is defined by `--logformat` option
	EnvLogFormat = "ARGOCD_LOG_FORMAT"
	// EnvLogFormatEnableTimestamp to indicate whether or not to enable the FullTimestamp option in log logs
	EnvLogFormatEnableTimestamp = "ARGOCD_LOG_FORMAT_ENABLE_TIMESTAMP"
	// EnvLogLevel log level that is defined by `--loglevel` option
	EnvLogLevel = "ARGOCD_LOG_LEVEL"
	// EnvMaxCookieNumber max number of chunks a cookie can be broken into
	EnvMaxCookieNumber = "ARGOCD_MAX_COOKIE_NUMBER"
	// EnvPluginSockFilePath allows to override the pluginSockFilePath for repo server and cmp server
	EnvPluginSockFilePath = "ARGOCD_PLUGINSOCKFILEPATH"
)
```

- Add in the file `util/log/logrus.go` the necessary changes to enable or disable the `FullTimeStamp` option based in the value of environment variable `ARGOCD_LOG_FORMAT_ENABLE_TIMESTAMP` through a local function, also it could be interesting adding some an extra local function to check the value of the variable `FORCE_LOG_COLORS`:

```
package log

import (
	"fmt"
	"os"
	"strings"

	adapter "github.com/bombsimon/logrusr"
	"github.com/go-logr/logr"
	"github.com/sirupsen/logrus"

	"github.com/argoproj/argo-cd/v2/common"
)

const (
	JsonFormat = "json"
	TextFormat = "text"
)

func NewLogrusLogger(fieldLogger logrus.FieldLogger) logr.Logger {
	return adapter.New(fieldLogger, adapter.WithFormatter(func(val interface{}) string {
		return fmt.Sprintf("%v", val)
	}))
}

// NewWithCurrentConfig create logrus logger by using current configuration
func NewWithCurrentConfig() *logrus.Logger {
	l := logrus.New()
	l.SetFormatter(CreateFormatter(os.Getenv(common.EnvLogFormat)))
	l.SetLevel(createLogLevel())
	return l
}

// CreateFormatter create logrus formatter by string
func CreateFormatter(logFormat string) logrus.Formatter {
	var formatType logrus.Formatter
	switch strings.ToLower(logFormat) {
	case JsonFormat:
		formatType = &logrus.JSONFormatter{}
	case TextFormat:
		formatType = &logrus.TextFormatter{
			ForceColors:   checkForceLogColors(),
			FullTimestamp: checkEnableFullTimestamp(),
		}
	default:
		formatType = &logrus.TextFormatter{
			FullTimestamp: checkEnableFullTimestamp(),
		}
	}

	return formatType
}

func checkForceLogColors() bool {
	return strings.ToLower(os.Getenv("FORCE_LOG_COLORS")) == "1"
}

func checkEnableFullTimestamp() bool {
	return strings.ToLower(os.Getenv(common.EnvLogFormatEnableTimestamp)) == "1"
}

func createLogLevel() logrus.Level {
	level, err := logrus.ParseLevel(os.Getenv(common.EnvLogLevel))
	if err != nil {
		level = logrus.InfoLevel
	}
	return level
}

### PR DESCRIPTION
**Pull request by SavelyKalinov**
Signed-off-by: Savely Kalinov [savelykalinov@gmail.com](mailto:savelykalinov@gmail.com)
This PR is a continuation of the @xoanmm`s work: https://github.com/argoproj/argo-cd/pull/8525.
Closes #8523

Checklist:

* [x] @xoanmm created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/8523) and discussed it with the community.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

**Review by todaywasawesome — APPROVED** [review]
Good to go

File: util/log/logrus_test.go [hunk 1]
```diff
@@ -23,6 +26,16 @@ func TestCreateFormatter(t *testing.T) {
 			result := CreateFormatter("text")
 			assert.Equal(t, &logrus.TextFormatter{}, result)
 		})
+		t.Run(fmt.Sprintf("%s == 1", common.EnvLogFormatEnableFullTimestamp), func(t *testing.T) {
+            os.Setenv(common.EnvLogFormatEnableFullTimestamp, "1")
```
Comment (pasha-codefresh) [inline util/log/logrus_test.go:30]:
```suggestion
            t.Setenv(common.EnvLogFormatEnableFullTimestamp, "1")
```

File: common/common.go [hunk 2]
```diff
@@ -242,6 +242,8 @@ const (
 	EnvLogFormat = "ARGOCD_LOG_FORMAT"
 	// EnvLogLevel log level that is defined by `--loglevel` option
 	EnvLogLevel = "ARGOCD_LOG_LEVEL"
+	// EnvLogFormatEnableFullTimestamp is used to indicate whether or not to enable the FullTimestamp option in logs
+	EnvLogFormatEnableFullTimestamp = "ARGOCD_LOG_FORMAT_ENABLE_FULL_TIMESTAMP"
```
Comment (todaywasawesome) [inline common/common.go:246]:
```suggestion
	// EnvLogFormatEnableFullTimestamp enables the FullTimestamp option in logs
	EnvLogFormatEnableFullTimestamp = "ARGOCD_LOG_FORMAT_ENABLE_FULL_TIMESTAMP"
```

Keep consistency in style

File: util/log/logrus_test.go [hunk 3]
```diff
@@ -23,6 +26,16 @@ func TestCreateFormatter(t *testing.T) {
 			result := CreateFormatter("text")
 			assert.Equal(t, &logrus.TextFormatter{}, result)
 		})
+		t.Run(fmt.Sprintf("%s == 1", common.EnvLogFormatEnableFullTimestamp), func(t *testing.T) {
+            t.Setenv(common.EnvLogFormatEnableFullTimestamp, "1")
+            result := CreateFormatter("text")
+            assert.Equal(t, &logrus.TextFormatter{FullTimestamp: true}, result)
+        })
+        t.Run(fmt.Sprintf("%s != 1", common.EnvLogFormatEnableFullTimestamp), func(t *testing.T) {
+            os.Setenv(common.EnvLogFormatEnableFullTimestamp, "0")
```
Comment (pasha-codefresh) [inline util/log/logrus_test.go:35]:
```suggestion
            t.Setenv(common.EnvLogFormatEnableFullTimestamp, "0")
```

**Comment by pasha-codefresh** [comment]
@SavelyKalinov oh, i am sorry , i didnt know that you still actively monitoring it, thats exactly what i am trying figure out :)

File: util/log/logrus_test.go [hunk 4]
```diff
@@ -1,10 +1,13 @@
 package log
 
 import (
+    "fmt"
+    "os"
```
Comment (pasha-codefresh) [inline util/log/logrus_test.go:5]:
```suggestion
```

**Comment by pasha-codefresh** [comment]
I probably found it, you used os package in the test, which requires mandatory error handling, so i changed it to t.setenv, actually this change makes it consistent with all other tests

**Comment by pasha-codefresh** [comment]
@SavelyKalinov could you please just open your IDE and format the code inside logrus_test.go, it is very hard to do from github :) either give me access to your fork, so i can help

**Comment by SavelyKalinov** [comment]
I see the same errors as last time:
```
Error: /usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/structured-merge-diff/v4@v4.4.1/fieldpath/pathelementmap.go: Cannot open: File exists
Error: /usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/structured-merge-diff/v4@v4.4.1/fieldpath/fromvalue_test.go: Cannot open: File exists
Error: /usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/structured-merge-diff/v4@v4.4.1/fieldpath/set.go: Cannot open: File exists
```
What's wrong with them? It looks like info rather than an error.

**Comment by pasha-codefresh** [comment]
> I see the same errors as last time:
> 
> ```
> Error: /usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/structured-merge-diff/v4@v4.4.1/fieldpath/pathelementmap.go: Cannot open: File exists
> Error: /usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/structured-merge-diff/v4@v4.4.1/fieldpath/fromvalue_test.go: Cannot open: File exists
> Error: /usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/structured-merge-diff/v4@v4.4.1/fieldpath/set.go: Cannot open: File exists
> ```
> 
> What's wrong with them? It looks like info rather than an error.

I think it is not relevant

**Comment by pasha-codefresh** [comment]
So, next time you should do 
1. make install-lint-tools
2. make lint-local

It will autofix linter issues for you

**Review by pasha-codefresh — APPROVED** [review]

**Pull request merged**